File: src/main/java/org/jukeboxmc/command/internal/GameRuleCommand.java
package org.jukeboxmc.command.internal;

import org.apache.commons.lang3.math.NumberUtils;
import org.cloudburstmc.protocol.bedrock.data.command.CommandParamType;
import org.jukeboxmc.command.Command;
import org.jukeboxmc.command.CommandData;
import org.jukeboxmc.command.CommandParameter;
import org.jukeboxmc.command.CommandSender;
import org.jukeboxmc.command.annotation.Description;
import org.jukeboxmc.command.annotation.Name;
import org.jukeboxmc.command.annotation.Permission;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.world.gamerule.GameRule;

import java.util.Arrays;
import java.util.List;

/**
 * @author LucGamesYT
 * @version 1.0
 */
@Name ( "gamerule" )
@Description ( "Set a gamerule value" )
@Permission ( "jukeboxmc.command.gamerule" )
public class GameRuleCommand extends Command {

    public GameRuleCommand() {
        super( CommandData.builder()
                .setParameters( new CommandParameter[]{
                                new CommandParameter( "rule", Arrays.stream( GameRule.values() ).map( gameRule -> gameRule.getIdentifier().toLowerCase() ).toList(), false ),
                                new CommandParameter( "value", List.of( "true", "false" ), false )
                        },
                        new CommandParameter[]{
                                new CommandParameter( "rule", Arrays.stream( GameRule.values() ).map( gameRule -> gameRule.getIdentifier().toLowerCase() ).toList(), false ),
                                new CommandParameter( "value", CommandParamType.INT, false ),

                        }
                ).build() );
    }

    @Override
    public void execute( CommandSender commandSender, String command, String[] args ) {
        if ( commandSender instanceof Player player ) {
            if ( args.length == 2 ) {
                String gameRuleValue = args[0];
                GameRule gameRule = GameRule.fromIdentifier( gameRuleValue );
                if ( gameRule != null ) {
                    if ( gameRule.getType().equals( GameRule.Type.INT ) ) {
                        if ( NumberUtils.isCreatable( args[1] ) ) {
                            int value = Integer.parseInt( args[1] );
                            player.getWorld().setGameRule( gameRule, value );
                            player.sendMessage( "Gamerule " + gameRule.getIdentifier() + " has been updated to " + value );
                        } else {
                            player.sendMessage( "§cValue must be a integer." );
                        }
                    } else {
                        String boolValue = args[1];
                        if ( boolValue.equalsIgnoreCase( "true" ) || boolValue.equalsIgnoreCase( "false" ) ) {
                            boolean value = Boolean.parseBoolean( boolValue );
                            player.getWorld().setGameRule( gameRule, value );
                            player.sendMessage( "Gamerule " + gameRule.getIdentifier() + " has been updated to " + value );
                        } else {
                            player.sendMessage( "§cValue must be a boolean." );
                        }
                    }
                } else {
                    player.sendMessage( "§cGameRule not found." );
                }
            }
        } else {
            commandSender.sendMessage( "§cYou must be a player." );
        }
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/command/internal/GameRuleCommand.kt
package org.jukeboxmc.server.command.internal

import org.apache.commons.lang3.math.NumberUtils
import org.jukeboxmc.api.command.Command
import org.jukeboxmc.api.command.CommandSender
import org.jukeboxmc.api.command.ParameterType
import org.jukeboxmc.api.command.annotation.*
import org.jukeboxmc.api.player.Player
import org.jukeboxmc.api.world.gamerule.GameRule

@Name("gamerule")
@Description("Set a gamerule value")
@Permission("jukeboxmc.command.gamerule")
@Parameters(parameter = [
    Parameter("gamerule", enumValues = arrayOf(
            "commandblocksenabled", "commandblockoutput", "dodaylightcycle",
            "doentitydrops", "dofiretick", "doinsomnia", "doimmediaterespawn", "domobloot", "domobspawning",
            "dotiledrops", "doweathercycle", "drowningdamage", "falldamage", "firedamage", "keepinventory",
            "maxcommandchainlength", "mobgriefing", "naturalregeneration", "pvp", "randomtickspeed", "sendcommandfeedback",
            "showcoordinates", "showdeathmessages", "spawnradius", "tntexplodes", "showtags", "freezedamage",
            "respawnblocksexplode", "showbordereffect", "functioncommandlimit", "recipesunlock", "dolimitedcrafting",
            "playerssleepingpercentage",
    ), optional = false),
    Parameter("value", enumValues = arrayOf("true", "false"), optional = false)
])
@Parameters(parameter = [
    Parameter("gamerule", enumValues = arrayOf(
            "commandblocksenabled", "commandblockoutput", "dodaylightcycle",
            "doentitydrops", "dofiretick", "doinsomnia", "doimmediaterespawn", "domobloot", "domobspawning",
            "dotiledrops", "doweathercycle", "drowningdamage", "falldamage", "firedamage", "keepinventory",
            "maxcommandchainlength", "mobgriefing", "naturalregeneration", "pvp", "randomtickspeed", "sendcommandfeedback",
            "showcoordinates", "showdeathmessages", "spawnradius", "tntexplodes", "showtags", "freezedamage",
            "respawnblocksexplode", "showbordereffect", "functioncommandlimit", "recipesunlock", "dolimitedcrafting",
            "playerssleepingpercentage",
    ), optional = false),
    Parameter("value", ParameterType.INT, optional = false)
])
class GameRuleCommand : Command {

    override fun execute(commandSender: CommandSender, command: String, args: Array<String>) {
        if (commandSender is Player) {
            if (args.size == 1) {
                val gameRuleValue = args[0]
                val gameRule = GameRule.fromIdentifier(gameRuleValue)
                if (gameRule != null) {
                    if (gameRule.type == GameRule.Type.INT) {
                        val value = commandSender.getWorld().getGameRule<Int>(gameRule)
                        commandSender.sendMessage(gameRule.identifier + " = " + value)
                    } else {
                        val value = commandSender.getWorld().getGameRule<Boolean>(gameRule)
                        commandSender.sendMessage(gameRule.identifier + " = " + value)
                    }
                } else {
                    commandSender.sendMessage("§cGameRule not found.")
                }
            } else if (args.size == 2) {
                val gameRuleValue = args[0]
                val gameRule = GameRule.fromIdentifier(gameRuleValue)
                if (gameRule != null) {
                    if (gameRule.type == GameRule.Type.INT) {
                        if (NumberUtils.isCreatable(args[1])) {
                            val value = args[1].toInt()
                            commandSender.getWorld().setGameRule(gameRule, value)
                            commandSender.sendMessage("Gamerule " + gameRule.identifier + " has been updated to " + value)
                        } else {
                            commandSender.sendMessage("§cValue must be a integer.")
                        }
                    } else {
                        val boolValue = args[1]
                        if (boolValue.equals("true", ignoreCase = true) || boolValue.equals("false", ignoreCase = true)) {
                            val value = boolValue.toBoolean()
                            commandSender.getWorld().setGameRule(gameRule, value)
                            commandSender.sendMessage("Gamerule " + gameRule.identifier + " has been updated to " + value)
                        } else {
                            commandSender.sendMessage("§cValue must be a boolean.")
                        }
                    }
                } else {
                    commandSender.sendMessage("§cGameRule not found.")
                }
            }
        } else {
            commandSender.sendMessage("§cYou must be a player.")
        }
    }
}